File: packages/realtime-protocol/src/limits.ts
import { z } from "zod";

export interface RealtimeLimitsV1 {
  readonly maxFrameBytes: number;
  readonly maxYjsUpdateBytes: number;
  readonly maxYjsStateVectorBytes: number;
  readonly maxMutationUpdateBytes: number;
  readonly maxRelativePositionBytes: number;
  readonly maxStringBytes: number;
  readonly maxAssistanceAcceptedDiffBytes: number;
}

export const DEFAULT_REALTIME_LIMITS_V1: RealtimeLimitsV1 = Object.freeze({
  maxFrameBytes: 4 * 1024 * 1024,
  maxYjsUpdateBytes: 2 * 1024 * 1024,
  maxYjsStateVectorBytes: 256 * 1024,
  maxMutationUpdateBytes: 2 * 1024 * 1024,
  maxRelativePositionBytes: 4 * 1024,
  maxStringBytes: 4 * 1024,
  maxAssistanceAcceptedDiffBytes: 1024 * 1024,
});

const RealtimeLimitsV1Schema = z.object({
  maxFrameBytes: z.number().int().min(12).max(0xffff_ffff),
  maxYjsUpdateBytes: z.number().int().positive().max(0xffff_ffff),
  maxYjsStateVectorBytes: z.number().int().positive().max(0xffff_ffff),
  maxMutationUpdateBytes: z.number().int().positive().max(0xffff_ffff),
  maxRelativePositionBytes: z.number().int().positive().max(0xffff_ffff),
  maxStringBytes: z.number().int().positive().max(0xffff),
  maxAssistanceAcceptedDiffBytes: z.number().int().nonnegative().max(0xffff_ffff),
}).strict();

export function validateRealtimeLimitsV1(limits: RealtimeLimitsV1): RealtimeLimitsV1 {
  const parsed = RealtimeLimitsV1Schema.parse(limits);
  for (const [name, maximum] of [
    ["maxYjsUpdateBytes", parsed.maxYjsUpdateBytes],
    ["maxYjsStateVectorBytes", parsed.maxYjsStateVectorBytes],
    ["maxMutationUpdateBytes", parsed.maxMutationUpdateBytes],
    ["maxRelativePositionBytes", parsed.maxRelativePositionBytes],
    ["maxStringBytes", parsed.maxStringBytes],
    ["maxAssistanceAcceptedDiffBytes", parsed.maxAssistanceAcceptedDiffBytes],
  ] as const) {
    if (maximum + 32 > parsed.maxFrameBytes) {
      throw new Error(`${name} must fit inside maxFrameBytes`);
    }
  }
  return parsed;
}

File: crates/oleafly-realtime-protocol/src/limits.rs
use crate::{ContractError, Result};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct RealtimeLimitsV1 {
    pub max_frame_bytes: usize,
    pub max_yjs_update_bytes: usize,
    pub max_yjs_state_vector_bytes: usize,
    pub max_mutation_update_bytes: usize,
    pub max_relative_position_bytes: usize,
    pub max_string_bytes: usize,
    pub max_assistance_accepted_diff_bytes: usize,
}

impl Default for RealtimeLimitsV1 {
    fn default() -> Self {
        Self {
            max_frame_bytes: 4 * 1024 * 1024,
            max_yjs_update_bytes: 2 * 1024 * 1024,
            max_yjs_state_vector_bytes: 256 * 1024,
            max_mutation_update_bytes: 2 * 1024 * 1024,
            max_relative_position_bytes: 4 * 1024,
            max_string_bytes: 4 * 1024,
            max_assistance_accepted_diff_bytes: 1024 * 1024,
        }
    }
}

impl RealtimeLimitsV1 {
    pub fn validate(self) -> Result<Self> {
        if self.max_frame_bytes < crate::FRAME_HEADER_LENGTH
            || self.max_frame_bytes > u32::MAX as usize
        {
            return Err(ContractError::InvalidFrame(
                "max_frame_bytes is outside the supported range".to_owned(),
            ));
        }
        for (name, value) in [
            ("max_yjs_update_bytes", self.max_yjs_update_bytes),
            (
                "max_yjs_state_vector_bytes",
                self.max_yjs_state_vector_bytes,
            ),
            ("max_mutation_update_bytes", self.max_mutation_update_bytes),
            (
                "max_relative_position_bytes",
                self.max_relative_position_bytes,
            ),
            ("max_string_bytes", self.max_string_bytes),
            (
                "max_assistance_accepted_diff_bytes",
                self.max_assistance_accepted_diff_bytes,
            ),
        ] {
            if value == 0 || value.saturating_add(32) > self.max_frame_bytes {
                return Err(ContractError::InvalidFrame(format!(
                    "{name} must fit inside max_frame_bytes"
                )));
            }
        }
        if self.max_string_bytes > u16::MAX as usize {
            return Err(ContractError::InvalidFrame(
                "max_string_bytes exceeds the wire string limit".to_owned(),
            ));
        }
        Ok(self)
    }
}
